src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala

```diff
@@ -101,6 +101,138 @@ case class ErgoTransaction(override val inputs: IndexedSeq[Input],
+          log.info(s"Tx verification failed: ${t.getMessage}")
+          (false, maxCost + 1)
+        case Success(result) =>
+          result
+      }
+
+    val currCost = addExact(currentTxCost, scriptCost)
+
+    validationBefore
+      // Check whether input box script interpreter raised exception
+      .validate(txScriptValidation, costTry.isSuccess && isCostValid, s"$id: #$inputIndex => $costTry")
+      // Check that cost of the transaction after checking the input becomes too big
+      .validate(bsBlockTransactionsCost, currCost <= maxCost, s"$id: cost exceeds limit after input #$inputIndex")
+      .map(c => addExact(c, scriptCost))
+  }
+
+  /**
+    * Helper method to verify an output of the transaction
+    *
+    * @param validationBefore - validation state before checking the output
+    * @param out - the output to check
+    * @param stateContext - context used in validation
+    * @return
+    */
+  private def verifyOutput(validationBefore: ValidationState[Long],
+                           out: ErgoBox,
+                           stateContext: ErgoStateContext): ValidationResult[Long] = {
+
+    val blockVersion = stateContext.blockVersion
+
+    validationBefore
+      .validate(txDust, out.value >= BoxUtils.minimalErgoAmount(out, stateContext.currentParameters), s"$id, output ${Algos.encode(out.id)}, ${out.value} >= ${BoxUtils.minimalErgoAmount(out, stateContext.currentParameters)}")
+      .validate(txFuture, out.creationHeight <= stateContext.currentHeight, s" ${out.creationHeight} <= ${stateContext.currentHeight} is not true, output id: $id: output $out")
+      .validate(txNegHeight, (blockVersion == 1) || out.creationHeight >= 0, s" ${out.creationHeight} >= 0 is not true, output id: $id: output $out")
+      .validate(txBoxSize, out.bytes.length <= MaxBoxSize.value, s"$id: output $out")
+      .validate(txBoxPropositionSize, out.propositionBytes.length <= MaxPropositionBytes.value, s"$id: output $out")
+  }
+
+  /**
+    * Check validity of transaction assets (access cost and preservation rules)
+    *
+    * @param validationBefore - validation state before checking the output
+    * @param outAssets - dictionary of assets in outputs
+    * @param outAssetsNum - number of assets in outputs
+    * @param boxesToSpend - input boxes
+    * @param stateContext - context used in validation
+    * @return
+    */
+  private def verifyAssets(validationBefore: ValidationState[Long],
+                           outAssets: Map[ByteBuffer, Long],
+                           outAssetsNum: Int,
+                           boxesToSpend: IndexedSeq[ErgoBox],
+                           stateContext: ErgoStateContext): ValidationResult[Long] = {
+    // Cost limit per block
+    val maxCost = stateContext.currentParameters.maxBlockCost
+
+    ErgoBoxAssetExtractor.extractAssets(boxesToSpend) match {
+      case Success((inAssets, inAssetsNum)) =>
+        lazy val newAssetId = ByteArrayWrapper(inputs.head.boxId)
```

Comment:
Using `ByteBuffer.wrap` here would be consistent with the asset-cost PR of mine

Reply:
This is a bug, as the `ErgoBoxAssetExtractor.extractAssets` returns `ByteBuffer`, so you need to do `ByteBuffer.wrap` I think